specification.md

```diff
@@ -3880,6 +3903,286 @@ execution of the state, runtimes should not create a new instance of this workfl
+  compensatedBy: CancelPurchase
+  transition:
+    nextState: SomeNextWorkflowState
+- name: CancelPurchase
+  type: operation
+  usedForCompensation: true
+  actions:
+  - functionRef:
+      refName: CreditCustomerFunction
+      parameters:
+        customerid: "{{ $.purchase.customerid }}"
+        amount: "{{ $.purchase.amount }}"
+  - functionRef:
+      refName: SendPurchaseCancellationEmailFunction
+      parameters:
+        customerid: "{{ $.purchase.customerid }}"
+```
+</td>
+</tr>
+</table>
+
+In this example our "NewItemPurchase" [event state](#Event-state) waits for a "NewPurchase" event and then 
+debits the customer and sends them a purchase confirmation email. It defines that it's compensated by the 
+"CancelPurchase" [operation state](#Operation-state) which performs two actions, namely credits back the 
+purchase amount to customer and sends them a purchase cancellation email.
+
+#### Triggering Compensation
+
+As previously mentioned, compensation must be explicitly triggered by the workflows control-flow logic. 
+This can be done via [transition](#Transition-definition) and [end](#End-definition) definitions.
+
+Let's take a look at each:
+
+1. Compensation triggered on transition:
+
+<table>
+<tr>
+    <th>JSON</th>
+    <th>YAML</th>
+</tr>
+<tr>
+<td valign="top">
+
+```json
+{
+  "transition": {
+      "compensateBefore": true,
+      "nextState": "NextWorkflowState"
+  }
+}
+```
+</td>
+<td valign="top">
+
+```yaml
+transition:
+  compensateBefore: true
+  nextState: NextWorkflowState
+```
+</td>
+</tr>
+</table>
+
+Transitions can trigger compensations by specifying the `compensateBefore` property and setting it to `true`.
+This means that before the transition is executed (workflow continues its execution to the "NextWorkflowState" in this example),
+workflow compensation must be performed. 
+
+2. Compensation triggered by end definition:
+
+<table>
+<tr>
+    <th>JSON</th>
+    <th>YAML</th>
+</tr>
+<tr>
+<td valign="top">
+
+```json
+{
+  "end": {
+    "compensateBefore": true,
+    "kind":"default"
+  }
+}
+```
+</td>
+<td valign="top">
+
+```yaml
+end:
+  compensateBefore: true
+  kind: default
+```
+</td>
+</tr>
+</table>
+
+End definitions can trigger compensations by specifying the `compensateBefore` property and setting it to `true`.
+This means that before workflow finishes its execution workflow compensation must be performed. Note that 
+in case when the end definition has its "kind" set to "event", compensation must be performed before 
+producing the specified events and ending workflow execution.
+
+#### Compensation Execution Details
+
+Now that we have seen how to define and trigger compensation, we need to go into details on how compensation should be executed.
+Compensation is performed on all already successfully completed states (that define `compensatedBy`) in **reverse** order.
+Compensation is always done in sequential order, and should not be executed in parallel.
```

Comment:
when you trigger compensation (via control-flow logic) you are compensating the workflow that is "now". Multiple iterations of some work that was already done has created this "now" and being able to tell what prior work has changed ( go back in time) might not be possible. 
Does this answer this question? 

Another thing we will add is explicit compensation (referencing a specific state to be compensated rather that triggering workflow-wide compensation). Atm we support implicit compensation only.

Reply:
I think I understand what you're saying, but let's go work through an example.

I have a workflow with 2 main states and 2 corresponding compensation states (a & a', b & b'). 

The workflow bounces back and forth between a & b for several iterations before failing in state a.

In what order should the overall flow of execution occur?

1) a, b, a, b, a, a', b', a',  b', a'
2) a, b, a, b, a, a', b'
3) a, b, a, b, a, c', b', a'

The question is basically, do we compensation in reverse order of first execution of the states, or reverse of the last execution of the states?  Or in complete reverse order of all the executions of every state?

Reply:
1) if a workflow fails as you mentioned, that should be handled by error handling of state a, not compensation. compensation is always explicitly triggered by the control flow logic.
2) no matter how many times a state was executed, when compensation is triggered it will be marked as "executed" and its compensation handler will be triggered 1 time.
hth